x/staking/keeper/val_state_change.go

```diff
@@ -241,6 +250,63 @@ func (k Keeper) ApplyAndReturnValidatorSetUpdates(ctx context.Context) (updates
 		updates = append(updates, validator.ABCIValidatorUpdateZero())
 	}
 
+	// ApplyAndReturnValidatorSetUpdates checks if there is ConsPubKeyRotationHistory
+	// with ConsPubKeyRotationHistory.RotatedHeight == ctx.BlockHeight() and if so, generates 2 ValidatorUpdate,
+	// one for a remove validator and one for create new validator
+	historyObjects, err := k.GetBlockConsPubKeyRotationHistory(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, history := range historyObjects {
+		valAddr := history.OperatorAddress
+		if err != nil {
+			return nil, err
+		}
+
+		validator, err := k.GetValidator(ctx, valAddr)
+		if err != nil {
+			return nil, err
+		}
+
+		oldPk, ok := history.OldConsPubkey.GetCachedValue().(cryptotypes.PubKey)
+		if !ok {
+			return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidType, "Expecting cryptotypes.PubKey, got %T", oldPk)
+		}
+		oldTmPk, err := cryptocodec.ToCmtProtoPublicKey(oldPk)
+		if err != nil {
+			return nil, err
+		}
+
+		newPk, ok := history.NewConsPubkey.GetCachedValue().(cryptotypes.PubKey)
+		if !ok {
+			return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidType, "Expecting cryptotypes.PubKey, got %T", oldPk)
+		}
+		newTmPk, err := cryptocodec.ToCmtProtoPublicKey(newPk)
+		if err != nil {
+			return nil, err
+		}
+
+		if !(validator.Jailed || validator.Status != types.Bonded) {
```

Comment:
Add a comment here as this kind of checks can be confusing (and I think it might be wrong). This means that a validator update will be added/modified if the validator is not jailed (regardless of the bonding status); but if the validator is jailed only if the validator is bonded.
I don't think that being jailed AND bonded is possible tho.

Reply:
You cannot be jailed and bonded. As soon as you're jailed, you're put into the unbonding statue/queue. Agree with @facundomedica here. Let's (A) avoid `!(...)`  and (B) add a very good comment above explaining what we're doing 👍

Reply:
> This means that a validator update will be added/modified if the validator is not jailed (regardless of the bonding status)

 - No, if the validator is not jailed it'll definitely check the status of the validator (it can be unbonded or unbonding status)
 - if the validator is jailed it'll abort the updation (does'nt enter the loop)
 - if the validator jailed false and not in bonded status then also if condition doesn't satisfy.
```
Assuming the condition should as like below:
// validator cannot rotate it's keys if not bonded or jailed.
// - a validator can be unbonding state but jailed status false
// - a validator can be jailed and status can be unbonding
```